Blog: Add post on leveraging Katib for efficient RAG optimization. (#161)
kubeflow/blog: _posts/2025-02-21-katib-rag-optimization.md

---
toc: true
layout: post
categories: [katib]
description: "Leveraging Katib for efficient RAG optimization."
comments: true
title: "Optimizing RAG Pipelines with Katib: Hyperparameter Tuning for Better Retrieval & Generation"
author: "Varsha Prasad Narsing" (@varshaprasad96)
---

# Introduction

> As machine learning models become more sophisticated, optimising their performance remains a critical challenge. Kubeflow provides a robust component, [KATIB][Katib], designed for hyperparameter optimization and neural architecture search. As a part of the Kubeflow ecosystem, KATIB enables scalable, automated tuning of underlying machine learning models, reducing the manual effort required for parameter selection while improving model performance across diverse ML workflows.

Reviewer: ```suggestion
As artificial intelligence and machine learning models become more sophisticated, optimising their performance remains a critical challenge. Kubeflow provides a robust component, [KATIB][Katib], designed for hyperparameter optimization and neural architecture search. As a part of the Kubeflow ecosystem, KATIB enables scalable, automated tuning of underlying machine learning models, reducing the manual effort required for parameter selection while improving model performance across diverse ML workflows.
```

With Retrieval-Augmented Generation ([RAG][rag]) becoming an increasingly popular approach for improving search and retrieval quality, optimizing its parameters is essential to achieving high-quality results. RAG pipelines involve multiple hyperparameters that influence retrieval accuracy, hallucination reduction, and language generation quality. In this blog, we will explore how KATIB can be leveraged to fine-tune a RAG pipeline, ensuring optimal performance by systematically adjusting key hyperparameters.

# Let's Get Started!

## STEP 1: Setup

> Since compute resources are scarcer than a perfectly labeled dataset :), we’ll use a lightweight KinD cluster to run this example locally. Rest assured, this setup can seamlessly scale to larger clusters by increasing the dataset size and the number of hyperparameters to tune.

Reviewer: ```suggestion
Since compute resources are scarcer than a perfectly labeled dataset :), we’ll use a lightweight [Kind cluster (Kubernetes in Docker)](https://kind.sigs.k8s.io/) to run this example locally. Rest assured, this setup can seamlessly scale to larger clusters by increasing the dataset size and the number of hyperparameters to tune.
```

> To get started, we'll first install the KATIB controller in our cluster by following the steps outlined [here][katib_installation].

Reviewer: ```suggestion
To get started, we'll first install the KATIB controller in our cluster by following the steps outlined [in the documentation][katib_installation].
```

## STEP 2: Implementing RAG pipeline

> In this implementation, we use a retriever model to fetch relevant documents based on a query and a generator model to produce coherent text responses.

Reviewer: probably would be good to outline what a retriever model is somewhere

### Implementation Details:

1. Retriever: Sentence Transformer & FAISS Index
   * A SentenceTransformer model (paraphrase-MiniLM-L6-v2) encodes predefined documents into vector representations.

>    * FAISS is used to index these document embeddings and perform efficient similarity searches to retrieve the most relevant documents.

Reviewer: ```suggestion
   * [FAISS](https://ai.meta.com/tools/faiss/) is used to index these document embeddings and perform efficient similarity searches to retrieve the most relevant documents.
```

2. Generator: Pre-trained GPT-2 Model
   * A Hugging Face GPT-2 text generation pipeline (which can be replaced with any other model) is used to generate responses based on the retrieved documents. I chose GPT-2 for this example as it is lightweight enough to run on my local machine while still generating coherent responses.
3. Query Processing & Response Generation
   * When a query is submitted, the retriever encodes it and searches the FAISS index for the top-k most similar documents.

>    * These retrieved documents are concatenated to form a context, which is then passed to the GPT-2 model to generate a response.

Reviewer: ```suggestion
   * These retrieved documents are concatenated to form the input context, which is then passed to the GPT-2 model to generate a response.
```

# Simulated RAG pipeline (simplified for example)
def rag_pipeline(query, top_k, temperature):
    return rag_pipeline_execute(query, top_k=top_k, temperature=temperature)

# Evaluate BLEU score
def evaluate(query, ground_truth, top_k, temperature):
    # Get the RAG pipeline response
    response = rag_pipeline(query, top_k, temperature)

    # Tokenize the response and ground truth
    reference = [ground_truth.split()]  # Reference should be a list of tokens
    candidate = response.split()  # Candidate is the generated response tokens

    # Apply smoothing to the BLEU score
    smoothie = SmoothingFunction().method1  # Use method1 for smoothing
    bleu_score = sentence_bleu(reference, candidate, smoothing_function=smoothie)

    return bleu_score

if __name__ == "__main__":
    # Parse command-line arguments
    parser = argparse.ArgumentParser(description="Evaluate BLEU score for a query using RAG pipeline")
    parser.add_argument("--top_k", type=int, required=True, help="Number of top documents to retrieve")
    parser.add_argument("--temperature", type=float, required=True, help="Temperature for the generator")
    args = parser.parse_args()

>     #TODO: The quries and ground truth against which the BLEU score needs to be evaluated. They can be provided in the script below or loaded from an external volume. 

Reviewer: ```suggestion
    #TODO: The queries and ground truth against which the BLEU score needs to be evaluated. They can be provided in the script below or loaded from an external volume. 
```

  metricsCollectorSpec: #  Specifies how Katib collects experiment results.
    collector:  
      kind: StdOut  #  Tells Katib to extract metrics from standard output logs.
  trialTemplate:
    primaryContainerName: training-container
    trialParameters:  #  Map hyperparameters (top_k and temperature) to their respective references in the job spec.
      - name: top_k
        description: Number of top documents to retrieve
        reference: top_k
      - name: temperature
        description: Temperature for text generation
        reference: temperature
    trialSpec:
      apiVersion: batch/v1
      kind: Job
      spec:
        template:
          spec:
            containers:
              - name: training-container
                image: rag-pipeline:latest
                command:
                  - "python"
                  - "/app/optimization-script.py"
                  - "--top_k=${trialParameters.top_k}"
                  - "--temperature=${trialParameters.temperature}"
                resources:
                  limits:
                    cpu: "1"
                    memory: "2Gi"
            restartPolicy: Never
```

> On applying this yaml on cluster, we can see our optimization script in action.

Reviewer: ```suggestion
After applying this yaml on the cluster, we can see our optimization script in action.
```

```commandline
kubectl get experiments.kubeflow.org -n kubeflow
NAME                    TYPE      STATUS   AGE
rag-tuning-experiment   Running   True     10m
```

We can also see the trials being run to search for the optimized parameter:

```commandline
kubectl get trials --all-namespaces
NAMESPACE   NAME                             TYPE      STATUS   AGE
kubeflow    rag-tuning-experiment-7wskq9b9   Running   True     10m
kubeflow    rag-tuning-experiment-cll6bt4z   Running   True     10m
kubeflow    rag-tuning-experiment-hzxrzq2t   Running   True     10m
```

> The list of completed trials and their results will be shown in the UI as below. Steps to access Katib UI are available [here][katib_ui].:  

Reviewer: ```suggestion
The list of completed trials and their results will be shown in the UI like below. Steps to access Katib UI are available [in the documentation][katib_ui].:  
```

![completed_runs](../images/2025-02-21-katib-rag-optimization/katib_experiment_run.jpeg)
![trial details](../images/2025-02-21-katib-rag-optimization/katib_ui.jpeg)

# Conclusion

In this experiment, we leveraged Kubeflow Katib to optimize a Retrieval-Augmented Generation (RAG) pipeline, systematically tuning key hyperparameters like top_k and temperature to enhance retrieval precision and generative response quality.

> For anyone working with RAG systems or hyperparameter optimization, Katib is a game-changer—enabling scalable, efficient, and intelligent tuning of machine learning models! In this experiment. Hope this helps you streamline hyperparameter tuning and unlock new efficiencies in your ML workflows!

Reviewer: ```suggestion
For anyone working with RAG systems or hyperparameter optimization, Katib is a powerful tool—enabling scalable, efficient, and intelligent tuning of machine learning models! We hope this tutorial helps you streamline hyperparameter tuning and unlock new efficiencies in your ML workflows!
```

---

> toc: true

Author: This did it, thanks @andreyvelich!

layout: post
categories: [ katib ]
description: "Leveraging Katib for efficient RAG optimization."
comments: true
title: "Optimizing RAG Pipelines with Katib: Hyperparameter Tuning for Better Retrieval & Generation"

> author: "Varsha Prasad Narsing" (@varshaprasad96)

Reviewer 2: ```suggestion
author: "Varsha Prasad Narsing (@varshaprasad96)"
```
I reproduced the cicd error locally, and this fixed it.

   # Apply smoothing to the BLEU score
   smoothie = SmoothingFunction().method1  # Use method1 for smoothing
   bleu_score = sentence_bleu(reference, candidate, smoothing_function=smoothie)

   return bleu_score

if __name__ == "__main__":
   # Parse command-line arguments
   parser = argparse.ArgumentParser(description="Evaluate BLEU score for a query using RAG pipeline")
   parser.add_argument("--top_k", type=int, required=True, help="Number of top documents to retrieve")
   parser.add_argument("--temperature", type=float, required=True, help="Temperature for the generator")
   args = parser.parse_args()

   # TODO: The queries and ground truth against which the BLEU score needs to be evaluated. 
   #  They can be provided in the script below or loaded from an external volume.
   query = ""
   ground_truth = " "

   # Call evaluate with arguments from the command line
   bleu_score = evaluate(query, ground_truth, args.top_k, args.temperature)
   print(f"BLEU={bleu_score}")
```

_Note_: Make sure to return the result in the format of `<parameter>=<value>`
for Katib's metrics collector to be able to utilize it. More ways to configure

> the output are available in [Kubeflow Metrics

Reviewer 3: ```suggestion
the output are available in [Katib Metrics
```

Collector][Katib_metrics_collector] guide.

The above scripts will be used to build our `rag-pipeline` image, that will be
specified in the `Experiment`.

> ## STEP 3: Run a KATIB Experiment

Reviewer 3: ```suggestion
## STEP 3: Run a Katib Experiment
```

To optimize the RAG pipeline’s hyperparameters, let's define an `Experiment` CR.
An experiment will define a single optimization run. More details on this API is
available in the [documentation][katib_api].

```yaml

> apiVersion: "kubeflow.org/v1beta1"

Reviewer 3: @varshaprasad96 I am wondering whether SDK submission would be easier, since we don't need to explain how to build Docker image from your RAG pipeline?
https://www.kubeflow.org/docs/components/katib/getting-started/#getting-started-with-katib-python-sdk

Reviewer 3: Can you use PyTorch as base_image and install required packages as follows:
```
def tune(
  base_image="docker.io/pytorch/pytorch:2.2.1-cuda12.1-cudnn8-runtime",
  packages_to_install=["rag",....],
)
```

Author: There are a few options I tried:
1. Use the generic base image as mentioned above, but embed the training_script in STEP 2 as a part of objective function - This worked, but required ample changes.
2. Build a custom image with training_script added to the container and pass it to `tune` API. In the objective function call in the script to run the optimization task.

Option 2 was much easier. I have added it as a note in the blog. Wdyt?

Reviewer 3: Docker image is fine, but it is not really what ML Users should do.
Please can you share your challenges to wrap  training script under the single function ?
As I can see your RAG pipeline is not super long.

Author: I see, that makes sense. As a developer, I found building a Docker image with the trainer script way much easier than restructuring the code. But that’s just my personal experience based on familiarity.

You're right—ML users shouldn't have to worry about writing or building Dockerfiles. I'll modify this section to use option 1 to better align with their workflow. Thanks for the feedback!

Reviewer 3: Thanks for sharing!
Yeah, we are trying to figure out how we can snapshot the user's file-system into Kubeflow Jobs, but this is still in-progress: https://github.com/kubeflow/trainer/issues/2347

cc @shravan-achar

kind: Experiment
metadata:
  name: rag-tuning-experiment
  namespace: kubeflow
spec:
  objective:
    type: maximize # Ensures that Katib tries to maximize the BLEU score.
    goal: 0.8 # Sets the desired BLEU score threshold.
    objectiveMetricName: BLEU # This should match with what we have provided in the script.
  algorithm:
    algorithmName: grid
  parameters:
    - name: top_k # Controls the number of retrieved documents.
      parameterType: int
      feasibleSpace:
        min: "1"
        max: "5"
        step: "1" # Adding a step for discrete search
    - name: temperature # Influences randomness in text generation.
      parameterType: double
      feasibleSpace:
        min: "0.5"
        max: "1.0"
        step: "0.1" # Adding a step for temperature
  metricsCollectorSpec: # Specifies how Katib collects experiment results.
    collector:

>       kind: StdOut # Tells Katib to extract metrics from standard output logs.

Reviewer 3: This is the default Metrics Collector, so we don't need to set it.

                command:
                  - "python"
                  - "/app/optimization-script.py"
                  - "--top_k=${trialParameters.top_k}"
                  - "--temperature=${trialParameters.temperature}"
                resources:
                  limits:
                    cpu: "1"
                    memory: "2Gi"
            restartPolicy: Never
```

After applying this yaml on the cluster, we can see our optimization script in
action.

```commandline
kubectl get experiments.kubeflow.org -n kubeflow
NAME                    TYPE      STATUS   AGE
rag-tuning-experiment   Running   True     10m
```

We can also see the trials being run to search for the optimized parameter:

```commandline
kubectl get trials --all-namespaces
NAMESPACE   NAME                             TYPE      STATUS   AGE
kubeflow    rag-tuning-experiment-7wskq9b9   Running   True     10m
kubeflow    rag-tuning-experiment-cll6bt4z   Running   True     10m
kubeflow    rag-tuning-experiment-hzxrzq2t   Running   True     10m
```

The list of completed trials and their results will be shown in the UI like
below. Steps to access Katib UI are available [in the documentation][katib_ui]:

> ![completed_runs](../images/2025-02-21-katib-rag-optimization/katib_experiment_run.jpeg)
> ![trial details](../images/2025-02-21-katib-rag-optimization/katib_ui.jpeg)

Reviewer 2: ```suggestion
![completed_runs](/images/2025-02-21-katib-rag-optimization/katib_experiment_run.jpeg)
![trial details](/images/2025-02-21-katib-rag-optimization/katib_ui.jpeg)
```

This wasn't rendered correctly from my end
![image](https://github.com/user-attachments/assets/c8a30f60-8b2d-4cf7-898e-17c8dd946dbd)

![image](https://github.com/user-attachments/assets/985e9369-da15-4b67-9ec1-a96a86d0cf0c)

   reference = [ground_truth.split()]  # Reference should be a list of tokens
   candidate = response.split()  # Candidate is the generated response tokens

   # Apply smoothing to the BLEU score
   smoothie = SmoothingFunction().method1  # Use method1 for smoothing
   bleu_score = sentence_bleu(reference, candidate, smoothing_function=smoothie)

   return bleu_score
```

## STEP 3: Run a Katib Experiment

To optimize the RAG pipeline’s hyperparameters, we need to submit an experiment to Katib, which will automatically 
search for the best configuration. We can do this using the [Katib SDK][Katib_SDK].

The first step would be to define an `objective` function which is what we will optimize. It runs the RAG pipeline with 
different hyperparameters and logs the BLEU score. 

```python
# Objective function for Katib
def objective(parameters):
    # TODO: Provide queries and ground truth directly here or load them dynamically from a file/external volume.
    query = ""  # Example: "Tell me about the Eiffel Tower."
    ground_truth = ""  # Example: "The Eiffel Tower is a famous landmark in Paris."

    # Run the RAG pipeline with hyperparameters from Katib

>     bleu_score = evaluate(query, ground_truth, top_k, temperature)

Reviewer 3: We should say that all imports and functions (e.g. `evaluate()`) should be defined under the `objective` function.
Similar to KFP Lightweight Python Component, the function must be self-contained: https://www.kubeflow.org/docs/components/pipelines/user-guides/components/lightweight-python-components/#python-function-requirements.

Users can also do something like this:
```python
def objective(parameters):
  from evaluate import eval
```

But in that case, evaluate module should be accessible in Katib Trials.

Author: I've restructured the content to talk about the objective function a clearly now
